Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,209 @@
+* If it's called when there already is CoreCLR in the process (loaded through the `hostfxr`, direct usage of `coreclr` is not supported), then the function determines if the specified runtime configuration is compatible with the existing runtime and frameworks. If it is, it returns a valid handle, otherwise it fails.
+
+It needs to be possible to call this function simultaneously from multiple threads at the same time as well.
+It also needs to be possible to call this function while there is an active host context created by `hostfxr_initialize_for_app` and running inside the `hostfxr_run_app`.
+
+*TODO: What is the indication that there is no runtime in the process yet?*
+
+
+### Inspect and modify host context
+
+#### Runtime properties
+``` C++
+int hostfxr_get_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    char_t * value_buffer,
+    size_t value_buffer_size,
+    size_t * value_buffer_used);
+
+int hostfxr_set_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    const char_t * value);
+```
+
+These functions allow the native host to inspect and modify runtime properties. After initialization the properties should contain everything which the hosting components calculates. This gives the native host the opportunity to inspect, modify and resolve any potential conflicts.
+* `host_context_handle` - the initialized host context.
+* `name` - the name of the runtime property to get/set. Must not be `nullptr`.
+* `value` - the value of the property to set. When set to `nullptr` the property is "unset" - removed from the runtime property collection.
+* `value_buffer` - buffer to receive the value of the property
+* `value_buffer_size` - the size of the `value_buffer` is `char_t` units.
+* `value_buffer_used` - when successful contains number of `char_t` units used in the buffer (including the null terminator). If the specified `value_buffer_size` was too small, the function returns `HostApiBufferTooSmall` and sets the `value_buffer_used` to the minimum required size.
+
+Trying to get a property which doesn't exist is an error and `hostfxr_get_runtime_property` will return appropriate error code.
+
+Setting properties is only supported when the host context was initialized without any runtime in the process. This is really a limitation of the runtime for which the runtime properties are immutable.
+
+*TODO: Can we support getting properties when there is preexisting runtime?*
+
+*TODO: Should there be a way to enumerate all the properties?*
+
+*TODO: Currently it's possible to have duplicate properties, hosting components take properties from `.runtimeconfig.json` and then append the calculated ones without any conflict resolution, potentially creating duplicates. It's important to not that if there are duplicates, CoreCLR will fail to start very early on.*
```

Comment:
That's actually a great idea! I think it's an acceptable change to fail on duplicates in the `hostpolicy` and don't wait on the runtime. The only observable behavior difference would be different failures
* Today depending on which property is duplicated and with which values the runtime may fail with different errors, but in the end it will always fail somehow.
* In the new world we would fail consistently the same way.

Let's actually do that regardless of the outcome of this document - it is a technically breaking change, but such a minor difference that we should be able to take it. And it makes the product much more predictable. And as you mention it let's us improve this in the future.

Reply:
I filed dotnet/core-setup#5529 for this.